src/numerics/petsc_matrix.C

```diff
@@ -1201,6 +1201,53 @@ void PetscMatrix<T>::add (const T a_in, const SparseMatrix<T> & X_in)
 }
 
 
+template <typename T>
+void PetscMatrix<T>::mat_mult (SparseMatrix<T> & X_in, SparseMatrix<T> & Y_out)
+{
+  libmesh_assert (this->initialized());
+
+   // sanity check.
+   libmesh_assert_equal_to (this->n(), X_in.m());
+   libmesh_assert_equal_to (this->m(), Y_out.m());
+   libmesh_assert_equal_to (X_in.n(), Y_out.n());
+
+   const PetscMatrix<T> * X = cast_ptr<const PetscMatrix<T> *> (&X_in);
+   PetscMatrix<T> * Y = cast_ptr<PetscMatrix<T> *> (&Y_out);
+
+   libmesh_assert (X);
+   libmesh_assert (Y);
+
+   PetscErrorCode ierr=0;
+
+   // the matrix from which we copy the values has to be assembled/closed
+   libmesh_assert(X->closed());
+
+   semiparallel_only();
+
+   ierr = MatMatMult(_mat, X->_mat, MAT_INITIAL_MATRIX, PETSC_DEFAULT, &Y->_mat);
+   LIBMESH_CHKERR(ierr);
+
+}
+
+template <typename T>
+void PetscMatrix<T>::add_sparse_matrix (const SparseMatrix<T> & spm,
```

Comment:
We already have
```
  /**
   * Compute \f$ A \leftarrow A + a*X \f$ for scalar \p a, matrix \p X.
   */
  virtual void add (const T a, const SparseMatrix<T> & X) = 0;
```
Would it be possible for you to use that instead? If not, then there's nothing here which is specific to PETSc so it would make more sense to add this new API at the SparseMatrix level.

Reply:
I do not think I will be able to use 

` virtual void add (const T a, const SparseMatrix<T> & X) = 0;`

because in my case X has smaller size than A, and I would like to add to certain rows and cols of A.

Reply:
OK, looks like we should also assert that `*this` has at least as many rows/cols as `spm` then? I also think you may find that this implementation is inefficient since it calls `add()` once per entry rather than calling `MatSetValues` with an entire row at a time. Therefore this could serve as a base class implementation, but @fdkong may be able to suggest an improvement for the `PetscMatrix` specialization.

One other small point is that I think you should declare `indices` and `values` outside the i-loop, no reason to create/destroy them `rows.size()` times.